### ISSUE TITLE
Simple-Sim: Feature Request: Introduce 3D Physically-Based Rendering Backend

### ISSUE THREAD

**Issue by arn-c0de** [arn-c0de/A-AIO-process-optimization-and-training#7]
The current dataset generation pipeline relies on 2D OpenCV rendering (`generator_2d.py`) to synthesize AOI-like ROI images.

While this works for basic classification, the images lack realistic depth cues, physically correct lighting, material interactions, shadows, reflections, and perspective — all critical for AOI in real PCB inspection.

**Proposal**: Add a 3D rendering backend (Blender + Cycles recommended) to generate photorealistic synthetic training data. Keep the 2D backend as default/fallback.

This should significantly improve **domain transfer** to real AOI camera images and generalization, especially for height- and lighting-dependent defects.

## Motivation

Current 2D limitations prevent models from learning physically meaningful features:

- No physically correct shadow casting
- No height-dependent illumination falloff / gradients
- No specular reflections (e.g. on shiny solder joints)
- No self-occlusion (lifted leads, raised components)
- No perspective distortion or lens effects
- No depth-of-field or focus blur variation

→ Models over-rely on 2D texture/edge patterns → poor real-world transfer.

**3D rendering enables learning AOI-relevant cues**:

- Realistic solder meniscus shape & curvature
- Shadow reveal of corner lifts / lifted leads
- Tombstoning via asymmetric lighting & contact gaps
- Solder bridges with proper coalescence & reflection
- Component height variations under dome/side lighting

Many recent works (and industrial tools like NVIDIA Omniverse Replicator) use physically-based 3D rendering for defect detection synthetic data — especially in electronics/PCB AOI — with strong gains in recall on height/volume-based defects.

## Proposed Solution

Introduce renderer-agnostic architecture via config flag:

```yaml
render_backend: opencv_2d      # default (current)
# render_backend: blender_3d   # new
```

### 1. New Module

Create `simple_sim/generator_3d.py`

Responsibilities:

- Load extended component profile
- Procedurally build parametric 3D geometry (pads, body, leads, solder fillets)
- Apply defect parameters (rotation, displacement, volume change, bridging geometry, etc.)
- Set up scene: materials (PBR), lighting (dome + directional), camera
- Render via Blender Python API (`bpy`) using Cycles
- Export: RGB image + JSON metadata (defect type, parameters, camera pose, etc.)

Recommended tools/libraries:

- Blender (headless mode via `blender --background`)
- `bpy` + BlenderProc (optional pipeline helper, very popular for CV synthetic data)
- Cycles for physically-based rendering

### 2. Updated Pipeline

```text
config → parameter sampling → geometry & defect build → scene setup → render (2D or 3D) → post-process → dataset store
```

### 3. Profile Extension (YAML)

Add 3D-relevant fields (example snippet):

```yaml
geometry:
  body_height_mm: 0.45
  pad_height_mm: 0.05
  lead_height_mm: 0.1
  lead_width_mm: 0.25

material:
  solder:
    roughness: 0.25
    metallic: 0.9
    ior: 1.4
  body:
    specular: 0.35
    base_color: [0.1, 0.1, 0.1]

lighting:
  dome_light:
    intensity: 1.2
    env_map: "dome.hdr"           # optional HDRI
  side_light:
    angle_deg: 35
    intensity: 0.8
  camera:
    focal_length_mm: 25
    sensor_size_mm: [4.8, 3.6]
    angle_deg: 8                  # tilt for realistic AOI view
```

### 4. Defect Mapping Table

| Defect (2D impl.)     | Current 2D Method          | Proposed 3D Method                              |
|-----------------------|----------------------------|-------------------------------------------------|
| MISALIGNED            | Pixel shift                | Spatial transform / translation                 |
| TOMBSTONE             | Rectangle tilt             | Actual rigid rotation + contact break           |
| SOLDER_BRIDGE         | Drawn blob                 | Merged / unioned solder mesh geometry           |
| CORNER_LIFT           | Label logic only           | Physically lifted corner + shadow gap           |
| INSUFFICIENT_SOLDER   | Smaller area               | Reduced fillet volume + concave meniscus        |
| EXCESS_SOLDER         | Larger area                | Increased fillet volume + possible spillover    |

## Expected Benefits

- **Data quality**: Physically consistent lighting, shadows, reflections → more realistic AOI appearance
- **Model performance**: Better real-world transfer, less overfitting to synthetic 2D edges, higher recall on height/geometry defects (target: +10–20% on real validation sets for lift/tombstone/bridge)
- **Future-proof**: Enables multi-view, lens distortion, HDR lighting, depth/normal map export, etc.

## Compatibility & Safety

- `opencv_2d` remains **default**
- Existing datasets unchanged (backwards compatible)
- Deterministic output via fixed random seeds
- No breaking changes to metadata schema (version if needed)

## Acceptance Criteria

- `render_backend: blender_3d` successfully generates a small dataset (≥100 images)
- Output is deterministic with fixed seed (`--seed 42`)
- Metadata schema compatible or explicitly versioned
- End-to-end training pipeline runs unchanged with new data
- Visual comparison shows realistic shadows, reflections, and height effects (attach example images in follow-up)
- Quantitative win: ≥ +10% recall / F1 on real AOI validation set for height-sensitive defects (corner lift, tombstone, bridging)

## Notes / References

- Directly resolves README MVP limitation: "3D rendering with Blender (MVP: 2D OpenCV only)"
- Related tools: BlenderProc (widely used for CV synthetic data), NVIDIA Replicator (industrial PCB defect examples)
- Effort estimate: medium-high (geometry modeling + Blender scripting), but high impact

**Comment by arn-c0de** [comment]
added in  0ef7d766dfc11a4533c66fdfb343903eb5c45967   -  status: in review
